### PR TITLE
[yugabyte/yugabyte-db#19315] Added support for streaming cql keyspaces

### PR DESCRIPTION
**Pull request by Sumukh-Phalgaonkar**
This PR adds the support requirred on the connector side to stream changes from CQL keyspaces. 

On the config side no new configs have been added. We will use the `database.dbname` field to specify keyspace name in case of ycql tables and namespace name in case of ysql tables. The `table.include.list` should contain a coma separated list of table names. The decision of whether the provided `dbname` and corresponding table list belongs to ycql or ysql is made on the basis of `streamid`.

The code path followed in case of ycql is for the most parts same as that of ysql. This PR introduces new value converter called `YugabyteDBCQLValueConverter` which is used for decoding of CQL specific column types.

### CODE REVIEW

File: src/main/java/io/debezium/connector/yugabytedb/YugabyteDBChangeRecordEmitter.java [hunk 1]
```diff
@@ -272,24 +272,27 @@ private Optional<DataCollectionSchema> newTable(TableId tableId) {
...
-            schema.refresh(connection, tableId,
-                           connectorConfig.skipRefreshSchemaOnMissingToastableData(),
-                           schema.getSchemaPBForTablet(tableId, tabletId), tabletId);
+            if (connectorConfig.qlType().equals("ysql")) {
+                schema.refresh(connection, tableId,
+                        connectorConfig.skipRefreshSchemaOnMissingToastableData(),
+                        schema.getSchemaPBForTablet(tableId, tabletId), tabletId);
+
+            } else {
+                schema.refresh(tableId, connectorConfig.skipRefreshSchemaOnMissingToastableData(),
```
Comment (vaibhav-yb) [inline src/main/java/io/debezium/connector/yugabytedb/YugabyteDBChangeRecordEmitter.java:292]: Put a comment before this explaining why are we not passing the `connection` object here or why is connection not required in this case.

File: src/main/java/io/debezium/connector/yugabytedb/YugabyteDBConnector.java [hunk 2]
```diff
@@ -95,34 +97,36 @@ public List<Map<String, String>> taskConfigs(int maxTasks) {
                     " available", maxTasks);
             return Collections.emptyList();
         }
-
         connection = new YugabyteDBConnection(yugabyteDBConnectorConfig.getJdbcConfig(), YugabyteDBConnection.CONNECTION_GENERAL);
+
```
Comment (vaibhav-yb) [inline src/main/java/io/debezium/connector/yugabytedb/YugabyteDBConnector.java:101]: nit: extra line

File: src/main/java/io/debezium/connector/yugabytedb/YugabyteDBConnectorConfig.java [hunk 3]
```diff
@@ -702,6 +702,14 @@ public JdbcConfiguration getJdbcConfig() {
             .withDescription(
                     "Whether or not to delete the logical replication stream when the connector finishes orderly" +
                             "By default the replication is kept so that on restart progress can resume from the last recorded location");
+        
+    public static final Field QL_TYPE = Field.create(DATABASE_CONFIG_PREFIX + "qltype")
+            .withDisplayName("The Query Language being used for tables to stream, in YugabyteDB (either ysql or ycql)")
+            .withType(Type.STRING)
+            .withImportance(Importance.MEDIUM)
+            .withDefault("ysql")
```
Comment (vaibhav-yb) [inline src/main/java/io/debezium/connector/yugabytedb/YugabyteDBConnectorConfig.java:710]: Add this default value to the top of the class to a final variable `DEFAULT_QL_TYPE`

File: src/main/java/io/debezium/connector/yugabytedb/YugabyteDBConnectorTask.java [hunk 4]
```diff
@@ -183,7 +197,7 @@ public ChangeEventSourceCoordinator<YBPartition, YugabyteDBOffsetContext> start(
                         }
                     });
 
-            final YugabyteDBEventDispatcher<TableId> dispatcher = new YugabyteDBEventDispatcher<>(
+            final YugabyteDBEventDispatcher<TableId> dispatcher = new YugabyteDBEventDispatcher<>(
```
Comment (vaibhav-yb) [inline src/main/java/io/debezium/connector/yugabytedb/YugabyteDBConnectorTask.java:200]: nit: extra space at the end of the line

File: src/main/java/io/debezium/connector/yugabytedb/YugabyteDBSchema.java [hunk 5]
```diff
@@ -248,8 +276,8 @@ public void readSchemaWithTablet(Tables tables, String databaseCatalog, String s
               .setPrimaryKeyNames(pkColumnNames)
               .setDefaultCharsetName(null /* default character set name */) // JDBC does not expose character sets
               .create();
-
-            String lookupKey = getLookupKey(tableId, tabletId);
+
```
Comment (vaibhav-yb) [inline src/main/java/io/debezium/connector/yugabytedb/YugabyteDBSchema.java:279]: nit: empty spaces

File: src/main/java/io/debezium/connector/yugabytedb/YugabyteDBSchema.java [hunk 6]
```diff
@@ -314,22 +342,28 @@ protected Optional<ColumnEditor> readTableColumn(CdcService.CDCSDKColumnInfoPB c
                                                      Tables.ColumnNameFilter columnFilter,
                                                      int position) {
         final String columnName = columnMetadata.getName();
-        final int oid = columnMetadata.getOid();
+        final int oid = columnMetadata.getOid();
```
Comment (vaibhav-yb) [inline src/main/java/io/debezium/connector/yugabytedb/YugabyteDBSchema.java:345]: nit: empty space at the end of line

File: src/test/java/io/debezium/connector/yugabytedb/YugabyteDBCQLTest.java [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+        verifyRecordCount(recordsCount);
+
+    }
+
+    private void verifyRecordCount(long recordsCount) {
+        waitAndFailIfCannotConsume(new ArrayList<>(), recordsCount);
+    }
+
+
+}
```
Comment (vaibhav-yb) [inline src/test/java/io/debezium/connector/yugabytedb/YugabyteDBCQLTest.java:86]: Does before image also work with CQL? If yes, can we add a basic test?

**Review by vaibhav-yb — CHANGES_REQUESTED** [review]

File: src/main/java/io/debezium/connector/yugabytedb/YBClientUtils.java [hunk 8]
```diff
@@ -230,7 +246,19 @@ public static YBClient getYbClient(YugabyteDBConnectorConfig connectorConfig) {
                                   .build();
     return new YBClient(asyncClient);
   }
-  
+
+  public static YBClient getYbClient(Configuration config) {
```
Comment (vaibhav-yb) [inline src/main/java/io/debezium/connector/yugabytedb/YBClientUtils.java:250]: Add a Javadoc comment for this method to explain why is this needed.

File: src/main/java/io/debezium/connector/yugabytedb/YBTableSchemaBuilder.java [hunk 9]
```diff
@@ -93,7 +97,8 @@ public YBTableSchemaBuilder(ValueConverterProvider valueConverterProvider, Schem
      * @param mappers the mapping functions for columns; may be null if none of the columns are to be mapped to different values
      * @return the table schema that can be used for sending rows of data for this table to Kafka Connect; never null
      */
-    public TableSchema create(String schemaPrefix, String envelopSchemaName, Table table, ColumnNameFilter filter, ColumnMappers mappers, KeyMapper keysMapper) {
+    public TableSchema create(String schemaPrefix, String envelopSchemaName, Table table, ColumnNameFilter filter,
+            ColumnMappers mappers, KeyMapper keysMapper) {
```
Comment (vaibhav-yb) [inline src/main/java/io/debezium/connector/yugabytedb/YBTableSchemaBuilder.java:101]: Indentation for this method should be corrected to follow YB styling.

File: src/main/java/io/debezium/connector/yugabytedb/YugabyteDBCQLValueConverter.java [hunk 10]
```diff
@@ -0,0 +1,555 @@
...
+                databaseCharset,
+                connectorConfig.getDecimalMode(),
+                connectorConfig.includeUnknownDatatypes(),
+                connectorConfig.hStoreHandlingMode(),
+                connectorConfig.intervalHandlingMode());
+    }
+
+    protected YugabyteDBCQLValueConverter(Charset databaseCharset, DecimalMode decimalMode, boolean includeUnknownDatatypes,
+                                       HStoreHandlingMode hStoreMode, IntervalHandlingMode intervalMode) {
+        // super(decimalMode, temporalPrecisionMode, defaultOffset, null, bigIntUnsignedMode, binaryMode);
```
Comment (vaibhav-yb) [inline src/main/java/io/debezium/connector/yugabytedb/YugabyteDBCQLValueConverter.java:172]: Do we need to remove this commented line now?

File: src/main/java/io/debezium/connector/yugabytedb/YugabyteDBConnector.java [hunk 11]
```diff
@@ -71,7 +73,7 @@ public Class<? extends Task> taskClass() {
     @Override
     public void start(Map<String, String> props) {
         this.props = props;
-        LOGGER.debug("Props " + props);
+        LOGGER.info("Props " + props);
```
Comment (vaibhav-yb) [inline src/main/java/io/debezium/connector/yugabytedb/YugabyteDBConnector.java:76]: I think we can change this back to debug, the properties are anyway printed once. With this we might end up printing all the properties twice. Please verify once and change back if that's the case.

File: src/main/java/io/debezium/connector/yugabytedb/YugabyteDBSchema.java [hunk 12]
```diff
@@ -249,7 +277,7 @@ public void readSchemaWithTablet(Tables tables, String databaseCatalog, String s
               .setDefaultCharsetName(null /* default character set name */) // JDBC does not expose character sets
               .create();
 
-            String lookupKey = getLookupKey(tableId, tabletId);
+            String lookupKey = getLookupKey(tableId, tabletId);;
```
Comment (vaibhav-yb) [inline src/main/java/io/debezium/connector/yugabytedb/YugabyteDBSchema.java:280]: nit: an extra semicolon got added

File: src/main/java/io/debezium/connector/yugabytedb/YugabyteDBSchema.java [hunk 13]
```diff
@@ -410,13 +469,22 @@ protected void refreshSchemaWithTablet(TableId id, String tabletId) {
...
-        TableSchemaBuilder schemaBuilder = getTableSchemaBuilder(config, valueConverter);
-        TableSchema schema = schemaBuilder.create(getSchemaPrefix(config.getLogicalName()), getEnvelopeSchemaName(table), table, config.getColumnFilter(), ColumnMappers.create(config), config.getKeyMapper());
 
+        TableSchema schema = schemaBuilder.create(getSchemaPrefix(config.getLogicalName()), getEnvelopeSchemaName(table), table, config.getColumnFilter(), ColumnMappers.create(config), config.getKeyMapper());
+      
         if (tableFilter.isIncluded(table.id())) {
             LOGGER.debug("Updating table schema with lookup key {}", lookupKey);
             tabletIdToTableSchema.put(lookupKey, schema);
+        } else {
+            LOGGER.warn("{} not included ",table.id());
```
Comment (vaibhav-yb) [inline src/main/java/io/debezium/connector/yugabytedb/YugabyteDBSchema.java:487]: Perhaps we can be more informative in the log and say "{} is not included in the table filter list"

Btw, what is the case where we can hit this else block?

File: src/main/java/io/debezium/connector/yugabytedb/YugabyteDBStreamingChangeEventSource.java [hunk 14]
```diff
@@ -565,13 +564,20 @@ protected void getChanges2(ChangeEventSourceContext context,
...
+                            if(connectorConfig.isYSQLDbType()) {
+                                // This is a hack to skip tables in case of colocated tables
+                                tempTid = YugabyteDBSchema.parseWithSchema(message.getTable(), pgSchemaNameInRecord);
 
-                            String pgSchemaNameInRecord = m.getPgschemaName();
-
-                            // This is a hack to skip tables in case of colocated tables
-                            TableId tempTid = YugabyteDBSchema.parseWithSchema(message.getTable(), pgSchemaNameInRecord);
+                            } else {
+                                tempTid = YugabyteDBSchema.parseWithKeyspace(message.getTable(), //TODO: Check if this breaks the above hack
```
Comment (vaibhav-yb) [inline src/main/java/io/debezium/connector/yugabytedb/YugabyteDBStreamingChangeEventSource.java:575]: I think we might not require the else block since colocation is only for YSQL tables.

File: src/main/java/io/debezium/connector/yugabytedb/YugabyteDBStreamingChangeEventSource.java [hunk 15]
```diff
@@ -692,9 +710,25 @@ else if (message.isDDLMessage()) {
...
+                                        dispatched = dispatched
+                                                && dispatcher.dispatchDataChangeEvent(part, tableId,
+                                                        new YugabyteDBChangeRecordEmitter(part, offsetContext, clock,
+                                                                connectorConfig,
+                                                                schema, connection, tableId, message,
+                                                                pgSchemaNameInRecord, tabletId,
+                                                                taskContext.isBeforeImageEnabled()));
+                                    } else {
+                                        dispatched = dispatched
+                                                && dispatcher.dispatchDataChangeEvent(part, tableId,
```
Comment (vaibhav-yb) [inline src/main/java/io/debezium/connector/yugabytedb/YugabyteDBStreamingChangeEventSource.java:725]: We seem to be doing the same thing in both the blocks, perhaps use a conditional operator to pass arguments?

```
 dispatched = dispatched
                                                && dispatcher.dispatchDataChangeEvent(part, tableId,
                                                        new YugabyteDBChangeRecordEmitter(part, offsetContext, clock,
                                                                connectorConfig,
                                                                schema, connection, tableId, message,
                                                                connectorConfig.isYSQLDbType() ? pgSchemaNameInRecord : tableId.catalog(),
                                                                tabletId,
                                                                taskContext.isBeforeImageEnabled()));
```

File: src/test/java/io/debezium/connector/yugabytedb/YugabyteDBCQLTest.java [hunk 16]
```diff
@@ -0,0 +1,161 @@
...
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicLong;
+
+import static org.junit.Assert.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.fail;
+
+public class YugabyteDBCQLTest extends YugabytedTestBase/*YugabyteDBContainerTestBase*/ {
```
Comment (vaibhav-yb) [inline src/test/java/io/debezium/connector/yugabytedb/YugabyteDBCQLTest.java:27]: nit: Change back the class before final push.

**Review by vaibhav-yb — CHANGES_REQUESTED** [review]

File: src/main/java/io/debezium/connector/yugabytedb/YugabyteDBSchema.java (hunk 13 above)
Comment (Sumukh-Phalgaonkar) [reply src/main/java/io/debezium/connector/yugabytedb/YugabyteDBSchema.java:488]: @vaibhav-yb Basically, since the table filters for cql and sql were different I was hitting an issue where Schema was not getting built for CQL. I found this particular log helpful while debugging so decided to keep it as a warning. Do you want me to remove this?

File: src/main/java/io/debezium/connector/yugabytedb/connection/pgproto/YbProtoReplicationMessage.java [hunk 17]
```diff
@@ -111,23 +113,39 @@ private List<ReplicationMessage.Column> transform(List<Common.DatumMessagePB> me
...
+                                return YbProtoReplicationMessage.this.getValue(columnName, type,
+                                        fullType, datum, connection, includeUnknownDatatypes);
+                            }
+                            @Override
+                            public String toString() {
+                                return datum.toString();
+                            }
+                        };
+                    }
+                    else {
```
Comment (vaibhav-yb) [inline src/main/java/io/debezium/connector/yugabytedb/connection/pgproto/YbProtoReplicationMessage.java:133]: Move this else back to previous line:
```
if {
  // content
} else {
  // else-content
}
```

File: src/main/java/io/debezium/connector/yugabytedb/connection/pgproto/YbProtoReplicationMessage.java [hunk 18]
```diff
@@ -111,23 +113,39 @@ private List<ReplicationMessage.Column> transform(List<Common.DatumMessagePB> me
...
+                        final Common.QLTypePB type = datum.getCqlType();
+                        final String fullType = typeInfo.map(CdcService.TypeInfo::getModifier).orElse(null);
+                        return new AbstractReplicationMessageColumn(columnName, type, fullType,
+                                typeInfo.map(CdcService.TypeInfo::getValueOptional).orElse(Boolean.FALSE), hasTypeMetadata()) {
+                            @Override
+                            public Object getValue(PgConnectionSupplier connection, boolean includeUnknownDatatypes) {
+                                return YbProtoReplicationMessage.this.getValue(columnName, type
+                                        , datum);
```
Comment (vaibhav-yb) [inline src/main/java/io/debezium/connector/yugabytedb/connection/pgproto/YbProtoReplicationMessage.java:141]: Formatting can be corrected here, the comma seems misplaced.

File: src/test/java/io/debezium/connector/yugabytedb/TestHelper.java [hunk 19]
```diff
@@ -380,9 +383,24 @@ public static Configuration.Builder getConfigBuilder(String namespaceName, Strin
                 .with(YugabyteDBConnectorConfig.STREAM_ID, dbStreamId);
     }
 
-    public static void setContainerHostPort(String host, int port) {
+    public static Configuration.Builder getConfigBuilderForCQL(String keyspaceName, String fullTableNameWithSchema, String dbStreamId) throws Exception {
+        return TestHelper.defaultConfig()
+                .with(YugabyteDBConnectorConfig.DATABASE_NAME, keyspaceName)
+                .with(YugabyteDBConnectorConfig.HOSTNAME,CONTAINER_YCQL_HOST)
+                .with(YugabyteDBConnectorConfig.PORT, CONTAINER_YCQL_PORT)
+                .with(YugabyteDBConnectorConfig.USER, "cassandra")
```
Comment (vaibhav-yb) [inline src/test/java/io/debezium/connector/yugabytedb/TestHelper.java:391]: We can add a static final variable named `DEFAULT_CASSANDRA_USER` with value `cassandra`

File: src/test/java/io/debezium/connector/yugabytedb/YugabyteDBDatatypesTest.java [hunk 20]
```diff
@@ -45,7 +45,6 @@ private void insertRecords(long numOfRowsToBeInserted) throws Exception {
             for (int i = 0; i < numOfRowsToBeInserted; i++) {
                 TestHelper.execute(String.format(formatInsertString, i));
             }
-
```
Comment (vaibhav-yb) [inline src/test/java/io/debezium/connector/yugabytedb/YugabyteDBDatatypesTest.java:48]: Unnecessary change.

File: src/main/java/io/debezium/connector/yugabytedb/YugabyteDBCQLValueConverter.java [hunk 21]
```diff
@@ -0,0 +1,1062 @@
...
+import io.debezium.jdbc.JdbcValueConverters.DecimalMode;
+import io.debezium.jdbc.ResultReceiver;
+import io.debezium.jdbc.TemporalPrecisionMode;
+import io.debezium.jdbc.ValueConversionCallback;
+import io.debezium.relational.Column;
+import io.debezium.relational.ValueConverter;
+import io.debezium.relational.ValueConverterProvider;
+import io.debezium.time.*;
+import io.debezium.util.NumberConversions;
+import io.debezium.util.Strings;
```
Comment (vaibhav-yb) [inline src/main/java/io/debezium/connector/yugabytedb/YugabyteDBCQLValueConverter.java:93]: Review all these imports once, I believe we are not using all of them, for example the postgres ones.

File: src/main/java/io/debezium/connector/yugabytedb/YugabyteDBCQLValueConverter.java
Comment (vaibhav-yb) [inline src/main/java/io/debezium/connector/yugabytedb/YugabyteDBCQLValueConverter.java:1]: Saw a few methods in this class which do not seem to be in use, if that is the case then we should remove them to avoid confusion.

If they are being used then some of them are throwing a `SQLException` which seems misleading as we are using a CQL converter here. Please review and remove/correct (whatever required).

File: src/main/java/io/debezium/connector/yugabytedb/YugabyteDBCQLValueConverter.java [hunk 22]
```diff
@@ -0,0 +1,1062 @@
...
+
+/**
+ * A provider of {@link ValueConverter}s and {@link SchemaBuilder}s for various YugabyteDB (ycql)
```
Comment (vaibhav-yb) [inline src/main/java/io/debezium/connector/yugabytedb/YugabyteDBCQLValueConverter.java:96]: *YCQL specific column types.

**Review by vaibhav-yb — CHANGES_REQUESTED** [review]
Mostly minor comments, but please take a look at the comments in the file `YugabyteDBCQLValueConverter` - I feel there are a lot of things there which we are not using.

File: src/main/java/io/debezium/connector/yugabytedb/YBClientUtils.java [hunk 23]
```diff
@@ -326,6 +360,39 @@ public static boolean isExplicitCheckpointingEnabled(YugabyteDBConnectorConfig c
               .equals(CdcService.CDCCheckpointType.EXPLICIT.name());
   }
 
+  public static Boolean isYSQLDatabaseType(String streamId, YBClient ybClient) {
+    GetDBStreamInfoResponse cdcStreamInfo = null;
+    ListNamespacesResponse resp = null;
+    try {
+      cdcStreamInfo = ybClient.getDBStreamInfo(streamId);
+      resp = ybClient.getNamespacesList();
```
[the remark this answers is not in the document]
Comment (Sumukh-Phalgaonkar) [reply src/main/java/io/debezium/connector/yugabytedb/YBClientUtils.java:368]: It was observed that these rpcs succeeded even when one node was available. When all the nodes are down, the RPCs are anyway retried in the yb-client. Also there's another diff which would fix the yb-client retries for `CDCErrorException`. Given all this, should we add retries from connector side over here?

File: src/main/java/io/debezium/connector/yugabytedb/connection/ReplicationMessageColumnValueResolver.java [hunk 24]
```diff
@@ -200,4 +202,42 @@ public static Object resolveValue(String columnName, YugabyteDBType type, String
...
+            case DECIMAL: return value.getValuepb().getDecimalValue();
+            case DATE: return value.getValuepb().getDateValue();
+            case VARINT: return value.getValuepb().getVarintValue();
+            case INET: return value.getValuepb().getInetaddressValue();
+            case LIST: return "list";
+            case MAP: return "map";
+            case SET: return "set";
+            case UUID: return value.getValuepb().getUuidValue();
+            case FROZEN: return "frozen";
+            case USER_DEFINED_TYPE: return "user_defined_type";
```
[the remark this answers is not in the document]
Comment (Sumukh-Phalgaonkar) [reply src/main/java/io/debezium/connector/yugabytedb/connection/ReplicationMessageColumnValueResolver.java:235]: Currently, we don't support these datatypes. The plan is to add support for them in coming versions.

**Review by vaibhav-yb — APPROVED** [review]
